src/ray/util/subreaper.cc

```diff
@@ -0,0 +1,235 @@
+
+using SignalHandlerFn = void (*)(const boost::system::error_code &, int);
+// Register a signal handler for the given signal.
+// The handler will be called with the signal_set and the error code.
+// After the handler is called, the signal will be re-registered.
+// The callback keeps a reference of the shared ptr to make sure it's not destroyed.
+void RegisterSignalHandlerLoop(std::shared_ptr<boost::asio::signal_set> signals,
+                               SignalHandlerFn handler) {
+  signals->async_wait(
+      [signals, handler](const boost::system::error_code &error, int signal_number) {
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+// Set this process as a subreaper.
+// Only works on Linux >= 3.4.
+void SetThisProcessAsSubreaper() {
+  if (prctl(PR_SET_CHILD_SUBREAPER, 1) == -1) {
+    RAY_LOG(FATAL) << "Failed to set this process as subreaper: " << strerror(errno);
+  }
+}
+
+// Kill all child processes that are not owned by this process.
+// It's done by checking the list of child processes and killing the ones that are not
+// created via Process::spawnvpe.
+//
+// TODO: Checking PIDs is not 100% reliable because of PID recycling. If we find issues
+// later due to this, we can use pidfd.
+void KillUnknownChildren() {
+  auto child_procs = GetAllProcsWithPpid(GetPID());
+
+  // Enumerating child procs is not supported on this platform.
+  if (!child_procs) {
+    RAY_LOG(FATAL) << "Killing leaked procs not supported on this platform. Only "
+                      "supports Linux >= 3.4";
+    return;
+  }
+  auto to_kill = KnownChildrenTracker::instance().listUnknownChildren(*child_procs);
+  for (auto pid : to_kill) {
+    RAY_LOG(INFO) << "Killing leaked child process " << pid;
+    auto error = KillProc(pid);
+    if (error && (*error)) {
+      RAY_LOG(WARNING) << "Failed to kill leaked child process " << pid << " with error "
+                       << error->message() << ", value = " << error->value();
+    }
+  }
+}
+
+// SIGCHLD handler that reaps dead children and kills unowned children.
+// Unowned children processes, or the process not created via Process::spawnvpe, can
+// happen when raylet becomes a subreaper, and a grandchild process is created then the
+// child process dies. The grandchild process becomes an orphan and is adopted by raylet
+// via Linux.
+//
+// T=0: raylet -> child -> grandchild
+// T=1: child receives SIGKILL. grandchild becomes an orphan and is reparented.
+// T=2: raylet -> child (zombie); raylet -> grandchild
+// T=3: raylet receives SIGCHLD for child
+// T=4: raylet reaps child, and finds grandchild is unowned child, killing it.
+//
+// CAVEAT: We may accidentally kill innocent subprocesses, if the direct child process
+// proposefully creates a grandchild process and exits. We need good test and audit to
+// make sure this doesn't happen.
+void SigchldHandlerKillOrphanSubprocesses(const boost::system::error_code &error,
```

Comment:
should we do this in sigchld? I feel like it is more robust & simple to just do it in a timer? (basically keep checking all unowned children and kill it?)